Add customization of Paired AMP URLs

westonruter opened the pull request:
Fixes #2204
Fixes #2062
Fixes #3357

This implements the customizability of a site's paired AMP URL structure, one of our most-frequently requested capabilities and one which will facilitate sites migrating from other AMP plugins. 

## Paired URL Structures

A new admin settings section is introduced for "Paired URL Structure" which allows the user to select from among four built-in paired URL structures:

* Query parameter, where `?amp=1` is used on all URLs. This is the default for new installs and it is the recommended structure since it does not conflict with any rewrite rules.
* Path suffix, where `/amp/` is used on all URLs. This is available due to popular demand (e.g. vanity reasons), although it offers no functional benefits over `?amp`.
* Legacy transitional, where `?amp` is used on all URLs. When upgrading the plugin from 2.0.x, a site in Transitional mode (or Reader mode with a Reader theme) will default to this structure.
* Legacy reader, where `/amp/` is used on all non-hierarchical posts and `?amp` is used on everything else. When upgrading the plugin from 2.0.x, a site in Transitional mode will default to this structure. (Note: This is the only structure that continues to support the `amp_pre_get_permalink` and `amp_get_permalink` filters.)

![image](https://user-images.githubusercontent.com/134745/100795935-b6ecfd00-33d4-11eb-9876-f49ef7bab045.png)

Each structure includes examples of how they will look:

![image](https://user-images.githubusercontent.com/134745/100795987-d08e4480-33d4-11eb-8444-8b439bc66625.png)

In the case of the path suffix and legacy reader paired URL structures (which both utilize `/amp/` path segment URL suffix), these structures will be unavailable if there is any entity on the site is currently using the `amp` slug (whether a post, term, user, post type, etc):

![image](https://user-images.githubusercontent.com/134745/100796208-219e3880-33d5-11eb-8fe8-d322ebfbe659.png)

When either the path suffix and legacy reader paired URL structures _are_ active, then a `wp_unique_post_slug` filter is used to prevent `amp` from being used a post slug, forcing it rather to be disambiguated like `amp-2`. (Again, the potential for conflicts is why the query parameter paired URL structure is recommended.)

In addition, users can implement their own paired URL strategy by extending the `AmpProject\AmpWP\PairedUrlStructure` class and supplying
this class name via the `amp_custom_paired_url_structure` filter. The subclass must implement two methods: `add_endpoint` and `remove_endpoint`.
A `has_endpoint` method can also be implemented, but the base class has a default implementation that uses `remove_endpoint`. For examples of how custom paired URL
structures can be implemented, see these mini plugins:

* [subdomain](https://gist.github.com/westonruter/67487d199f4cbc1968a30c60de121893) (e.g. `https://amp.example.com/foo/`)
* [path prefix](https://gist.github.com/westonruter/f881658cb1bc5a03cc432c8e2b89102a) (e.g. `https://example.com/amp/foo/`)

In order for the custom URL structure to work, the `amp_custom_paired_url_structure` filter must be added _before_ the `plugins_loaded` action at priority 7 (since the paired structure has to be set at this point to be used by `PairedRouting`). So the best bet is to add the filter at the top-level of your plugin and not inside of an action. For example:

```php
add_filter(
	'amp_custom_paired_url_structure',
	function () {
		require_once __DIR__ . '/PrefixUrlStructure.php';
		return PrefixUrlStructure::class;
	}
);
```

The same goes for filtering `amp_query_var`.

When a custom paired URL structure is used, all of the built-in structures are disabled and a notice is displayed:

![image](https://user-images.githubusercontent.com/134745/100796481-92455500-33d5-11eb-9f13-96f123d732cc.png)

## Paired Routing

Previously `?amp=1` paired URLs were the exclusive structure available when using a Reader theme because the `ReaderThemeLoader` service needed to rely on `$_GET['amp']` as a signal for whether to switch the theme in the request. It was not able to use the `/amp/` “rewrite endpoint” in WordPress because that depends on the `WP::parse_request` having been run, which is not available early enough (at `plugins_loaded`) as a signal for `ReaderThemeLoader`. In order to allow for _any_ paired URL structure to be used in any template mode (either Transitional, Reader theme, or legacy Reader), WordPress's rewrite rules are not used at all for detecting paired AMP URLs. Instead, the new `PairedRouting` service will examine the URL for matching against the currently-selected paired URL structure and store whether it is a paired AMP request. The paired AMP endpoint is then removed from the environment (e.g. `$_SERVER['REQUEST_URI']`) while `WP::parse_request()` runs so that all current rewrite rules will work as they do normally for non-AMP requests. The `amp` query var is added to the `WP::$query_vars`. Then at once the `WP::parse_request()` process is finished at the `parse_request` action, then the endpoint is re-added to `$_SERVER['REQUEST_URI']` as well as `$wp->request`.

To reiterate: Previously this plugin implemented the `/amp/` URL endpoint suffix by means of calling `add_rewrite_endpoint()` in the WordPress rewrite API. This worked adequately for singular post URLs, but it quickly became clear that it would not be suitable for use across all URLs on a given site. Even when `EP_ALL` is supplied to `add_rewrite_endpoint()` the reality is that _all_ rewrite rules do not get the endpoint: for example paged URLs , comment page URLs, and URLs which contain _other_ endpoints. For these reasons, the use of rewrite rules to implement paired URL structures was eliminated entirely in favor of inspecting the `REQUEST_URI` for the path segment or query parameter (or some other environment indicator). When WordPress finally gets to running `WP::parse_request()`, the AMP endpoint is stripped from the environment (i.e. `$_SERVER['REQUEST_URI']`) just before at the `do_parse_request` hook so that all of the site's rewrite rules will continue to match as if AMP was not being requested at all. Then after the rewrite rules have been matched at the `parse_request` action, the AMP endpoint is restored in the environment. The ability to identify paired AMP URLs before the `wp` action allows for all paired URL structures to be used even when using a Reader theme (which needs to know if the request is for an AMP URL at `plugins_loaded` in order to switch the theme). In addition to this allowing AMP endpoints to work with all URLs matched by a site's rewrite rules, it also means **there is no longer a need to flush rewrite rules when activating (or deactivating) the AMP plugin**.

The work on paired routing resulted paying down a good amount of tech debt namely be moving a lot of code from functions and the massive `AMP_Theme_Support` class into their respective.

## Paired Browsing

When Paired Browsing was first implemented (https://github.com/ampproject/amp-wp/pull/3656), the only paired URL structure applicable was the legacy Transitional mode (`?amp`). This meant it was trivial for the Paired Browsing JS code to obtain and identify an AMP URL from a non-AMP one and vice versa, since it meant just adding or removing the `amp` query parameter. With the introduction of customizable paired URL structures, this complicated matters, the biggest being that a paired URL structure may involve a separate _domain_ entirely (e.g. an `amp` subdomain). The fact that the AMP half of the Paired Browsing UI could be in an iframe that's on another domain meant that we could no longer rely on direct cross-window access in JS to synchronize the current URL and the scroll position: it was necessary to make use the `postMessage` API instead, in the same way that the Customizer does. So this PR includes this major refactor of Paired Browsing.

The refactor included extraction of a lot of code from the massive `AMP_Theme_Support` class into the new `PairedBrowsing` service.

## Post Previews

Similarly to the changes necessitated for Paired Browsing, the ability to view the AMP version of a Post Preview also needed to be changed. No longer could the AMP version of a post be obtained just by adding the `amp` query parameter. So now the AMP version of the post preview URL is obtained server-side and exported to the client for use in the Preview button.

Additionally, the editor plugin that introduces the AMP preview button is now skipped from registration when the site is in Standard mode. Query parameters which would be added to the preview URL (e.g. `preview_nonce` and `preview_id`) are copied on top of that initial AMP preview URL.

## Changes for `canonical` links

Previously when a non-AMP page lacked a `link[rel=canonical]` element, it would attempt to generate an actual “canonical” URL to supply for the required `link` added to AMP pages. In reality, in an AMP context this “canonical” link does not mean semantic canonicity but rather simply that it is “non-AMP”. Therefore, the logic has been changed so that on AMP pages, the `canonical` link is simply the current URL with the AMP endpoint removed when in a paired AMP mode (Transitional or Reader) and to just return the current URL when in Standard mode. In order to output actual semantically `canonical` links, add a plugin like [Yoast SEO](https://wordpress.org/plugins/wordpress-seo/) or advocate for core to output canonical links on non-singular responses (see [TRAC-18660](https://core.trac.wordpress.org/ticket/18660)).

## Mobile Redirection

If a site had mobile redirection enabled in Reader mode or Transitional mode and then they switched to Standard mode, the mobile redirect option would remain enabled. The `MobileRedirection` service was not checking whether Standard moe was active, resulting in needless hooks added for when a paired mode is active. This has been rectified in 2474e8d7386e3cf1a96cf4df8bf96ee5ac892726.

This PR also ensures that submitting a comment an AMP page will cause the redirected URL to be an AMP page, as opposed to the non-AMP page.

## 404 Handling

When accessing a URL results in a 404 and and the URL also ends in the `/amp/` URL suffix, the plugin will automatically try redirecting to `/?amp=1`. This ensures that sites that formerly used the `/amp/` URL suffix will not end up with broken URLs when switching. Fixes #2062.

## Old Slug Redirect Handling

When the 404 template was not enabled for AMP, attempting to access the AMP version of a URL which had the old post slug would end up getting redirected to the new permalink but _without_ the AMP endpoint. This was due to the `redirect_paired_amp_unavailable` running first to stip off the AMP endpoint when it detected the 404, before `wp_old_slug_redirect()` was able to run to redirect with the AMP endpoint persisted. So now in `redirect_paired_amp_unavailable` just before redirection happens we just-in-time call `wp_old_slug_redirect()` to make sure its redirection takes precedence. Fixes #3357.

## New APIs

### New Classes

* `\AmpProject\AmpWP\PairedRouting`: Service for routing users to and from paired AMP URLs.
* `\AmpProject\AmpWP\Admin\PairedBrowsing`: Service for the Paired Browsing app.
* `\AmpProject\AmpWP\PairedUrl`: Service for common manipulations for paired URLs, such as whether a given URL has a query var or endpoint suffix, or to remove/add them.
* `\AmpProject\AmpWP\PairedUrlStructure`: Abstract class which which a paired URL extends to describe how to make a URL paired, make it un-paired, and to determine if it is paired. 
* `\AmpProject\AmpWP\LegacyReaderUrlStructure`: Legacy Reader paired URL structure.
* `\AmpProject\AmpWP\LegacyTransitionalUrlStructure`: Legacy Transitional paired URL structure. 
* `\AmpProject\AmpWP\QueryVarUrlStructure`: Query Var paired URL structure.
* `\AmpProject\AmpWP\PathSuffixUrlStructure`: Path Suffix paired URL structure.

### Hook Changes

* The `amp_default_options` filter now passes the current `$options` as the second parameter. 
* A new `amp_validated_url_status_actions` filter allows services to inject links into the publish metabox on the validated URL screen.
* A new `amp_rest_options` filter is introduced for services to add (readonly) options to the AMP settings response. 
* A new `amp_rest_options_schema` filter is introduced for services to describe their own options.

### Deprecated Functions

* `amp_correct_query_when_is_front_page()`, now handled by `PairedRouting`.
* `amp_add_frontend_actions()`, now handled by `PairedRouting`.
* `amp_redirect_old_slug_to_new_url()`, now handled by `PairedRouting`.

### Plugin Suppression

The Plugin Suppression logic now defers initialization until `plugins_loaded` priority 9 as opposed to running much earlier at the minimum priority. This is because it needs to check if `has_endpoint` which depends on plugins possibly registering their own custom paired URL structure, say at `plugins_loaded` priority 0. (2d4dea46439bc562c2878a28ceec493cfed62980)

----

- [x] Fix post previews in block editor.
- [x] Ensure post previews in classic editor work.
- [x] Fix paired browsing.
- [x] Ensure the `/amp/` addition is added properly.
- [x] Flush rewrite rules when updating paired URL structure.
- [x] Add UI.
- [x] Fix being able to access static front page at `/amp/`.
- [x] Fix pagination of blog index page, from `https://wordpressdev.lndo.site/blog/amp/` to next page at  `https://wordpressdev.lndo.site/blog/amp/page/2/?amp=1`.
- [x] Make sure rewrite endpoints work for post type archives.
- [x] Disable /amp/ from being available when there is a post that has "amp" as the name. 
- [x] Automatically add -2 to the slug when using an endpoint.
- [x] 404 endpoint redirect.
- [x] Move more code from theme support, like ensure location.
- [ ] ~Consider removing legacy options.~ (To do in future.)
- [x] Ensure infinite address space is accounted for.
- [x] In standard mode, should `/amp/` invoke canonical redirect or should it strip the endpoint?
- [x] Move amphtml link addition to `PairedAmpRouting`.
- [x] Remove duplicate endpoint types, both `/amp/` and `?amp`.
- [x] Strip off `?amp=1` from URLs when AMP is canonical.
- [x] Restore `/amp/` to `$wp->request`.
- [x] Break up `PairedAmpRouting` into more services.
- [ ] Warn when `amp_custom_paired_url_structure` filter is added too late?
- [x] Add/update tests.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

westonruter committed:
Introduce context param for amp_add_paired_endpoint()

westonruter committed:
Undo adding context param to amp_add_paired_endpoint()

westonruter committed:
Introduce PairedAmpRouting service

westonruter committed:
Move amp_redirect_old_slug_to_new_url() to PairedAmpRouting

westonruter committed:
Move amp_correct_query_when_is_front_page() to PairedAmpRouting service

westonruter committed:
Add initial implementations of the 4 paired AMP permalink structures

westonruter committed:
Add filters for implementing custom paired AMP permalink structures

westonruter committed:
Fix ability to preview post using a custom paired AMP URL

westonruter committed:
Allow paired browsing to work for custom paired URLs

westonruter committed:
Use postMessage for navigation and scrolling

westonruter committed:
Remove unused vars

westonruter committed:
Remove obsolete ampPageHasErrors and improve isAmpDocument passing

westonruter committed:
Implement heartbeat check

westonruter commented on includes/templates/amp-paired-browsing.php:
```diff
@@ -43,16 +43,9 @@
 					</div>
 
 					<div class="dialog-buttons">
-						<button class="button exit">
-							<?php esc_html_e( 'Exit', 'amp' ); ?>
-						</button>
```
This can be restored if we capture link clicks in the preview and send a message to the parent containing the `href` being navigated to.

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -2386,10 +2401,11 @@ static function( $tag, $handle ) {
 	 */
 	public static function get_paired_browsing_url( $url = null ) {
```
All of the paired browsing logic should be moved into a new service.

westonruter commented on src/PairedAmpRouting.php:
```diff
@@ -0,0 +1,562 @@
+	 * @return string Paired AMP permalink structure.
+	 */
+	public function get_permalink_structure() {
+		$has_filters      = [
+			has_filter( 'amp_has_paired_endpoint' ),
+			has_filter( 'amp_add_paired_endpoint' ),
+			has_filter( 'amp_remove_paired_endpoint' ),
+		];
+		$has_filter_count = count( array_filter( $has_filters ) );
+		if ( 3 === $has_filter_count ) {
+			return self::PERMALINK_STRUCTURE_CUSTOM;
```
As noted by @schlessera, we may want to switch this around so that the filters do not _force_ the custom URL structure, but rather that the filters can merely allow the custom structure to be among the selectable options.

westonruter committed:
Restore exit link in modal

westonruter committed:
Fix initialization of iframes to populate document title

westonruter committed:
Eliminate dependence on window.name

westonruter committed:
Fix event delegation for navigate messages

westonruter committed:
Introduce PairedBrowsing service

westonruter committed:
Move option values to Options interface

westonruter committed:
Make amp_get_permalink() an alias for get_permalink() in Standard mode

westonruter committed:
Fix infinite recursion in get_options

westonruter committed:
Fix tests

westonruter committed:
Improve implementation of deprecated is_paired_available method

westonruter committed:
Rename 'permalink structure' to 'paired URL structure'

westonruter committed:
Add Paired URL Structure settings drawer

westonruter committed:
Flush rewrite rules when changing the paired URL structure

westonruter committed:
Improve formatting of paired URL examples

westonruter committed:
Add missing 'custom' to has_paired_url_structure method

westonruter committed:
Prevent appending /amp/ endpoint to URLs which already have rewrite endpoint(s)

westonruter committed:
Improve styling of paired URL structure items

westonruter commented:
Current UI looks like this:

![image](https://user-images.githubusercontent.com/134745/98502380-b5ca1500-2206-11eb-8c76-f462edf74146.png)

When a plugin has added a custom paired URL structure, it currently looks like this:

![image](https://user-images.githubusercontent.com/134745/98502524-0a6d9000-2207-11eb-80b9-a926fbfb1fee.png)

westonruter committed:
Use notice for custom paired URL structure; add examples

westonruter commented:
I've moved the Custom structure to a notice and added example URLs for each structure:

![image](https://user-images.githubusercontent.com/134745/98602697-ba89da00-2295-11eb-8e0b-2aa4290bee4c.png)

When Custom URLs are not being used:

![image](https://user-images.githubusercontent.com/134745/98602852-f329b380-2295-11eb-8efe-655363ec6cdd.png)

I'm not totally happy with how these options are laid out, but the information is there at least. /cc @jwold

westonruter committed:
Preserve AMP endpoint through redirect_canonical

westonruter committed:
Only add necessary filters when is not canonical

westonruter committed:
Fix malformed injection of amp endpoint in paginate_links()

westonruter committed:
Fix generating AMP URLs for paged, comment pages, and post previews

westonruter committed:
Fix regex for added_rewrite_endpoints_pattern

westonruter committed:
Make use of using_permalinks helper method

westonruter committed:
Check amp_is_request() once when adding filters

westonruter committed:
Eliminate use of add_rewrite_endpoint() for adding AMP suffix

westonruter commented:
Making ready for review to create build ZIPs.

pierlon commented:
> Making ready for review to create build ZIPs.

Maybe we could configure a slash command that triggers the build jobs on demand.

westonruter committed:
Account for slashing in REQUEST_URI

westonruter committed:
Tighten styling of Paired URL Structure drawer

westonruter commented:
@jwold I further tightened the styling of the drawer in 350393dcd.

No custom structure active:

Examples Collapsed | Examples Expanded
---------------------|---------------------
![image](https://user-images.githubusercontent.com/134745/99113985-f29a6100-25a4-11eb-92ff-0b85c264cf08.png) | ![image](https://user-images.githubusercontent.com/134745/99114008-fd54f600-25a4-11eb-8079-02a4cbc3b05c.png)

Custom structure active:

Examples Collapsed | Examples Expanded
---------------------|---------------------
![image](https://user-images.githubusercontent.com/134745/99113894-d4346580-25a4-11eb-896a-4af02b1bf5a5.png) | ![image](https://user-images.githubusercontent.com/134745/99113938-e31b1800-25a4-11eb-92d8-217006bad107.png)

Feedback welcome.

westonruter committed:
Introduce REST filters to move logic to service

westonruter committed:
Remove PAIRED_URL_STRUCTURE_CUSTOM from Option

westonruter committed:
Note redundancy of sanitize_options with enum

westonruter committed:
Remove obsolete filtering of paginate_links

westonruter committed:
Indicate the name of the theme/plugin which introduces the custom paired URL structure

westonruter committed:
Add notice on permalink settings screen for where to manage paried URL structure

westonruter committed:
Limit kses to what is actually needed

westonruter committed:
Add todos

westonruter committed:
Prevent selecting suffix endpoint when there are conflicts

westonruter committed:
Ensure that 'amp' query var is always registered

westonruter committed:
Avoid matching segments after /amp/ in URL paths

westonruter committed:
Use DI PairedAmpService in MobileRedirection

westonruter committed:
Fix tearDown by resetting $wp_the_query

westonruter committed:
Remove tests for removal of endpoint segments after /amp/

westonruter committed:
Fix OptionsRESTControllerTest by adding item

westonruter committed:
Fix MobileRedirection tests that should not have previously passed

westonruter commented on tests/php/src/MobileRedirectionTest.php:
```diff
@@ -151,7 +156,7 @@ public function test_redirect_on_transitional_and_not_available() {
 		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::TRANSITIONAL_MODE_SLUG );
 		AMP_Options_Manager::update_option( Option::ALL_TEMPLATES_SUPPORTED, false );
 		AMP_Options_Manager::update_option( Option::SUPPORTED_TEMPLATES, [ 'is_author' ] );
-		$this->go_to( amp_add_paired_endpoint( '/' ) );
```
Going to the AMP endpoint seems to have been a bug here. I'm unclear why it was passing tests before.

westonruter commented on tests/php/src/MobileRedirectionTest.php:
```diff
@@ -185,7 +190,8 @@ public function test_redirect_when_server_side_and_not_applicable() {
 		add_filter( 'amp_mobile_client_side_redirection', '__return_false' );
 		add_filter( 'amp_pre_is_mobile', '__return_false' );
 
-		$this->go_to( amp_add_paired_endpoint( '/' ) );
```
Going to the AMP endpoint seems to have been a bug here. I'm unclear why it was passing tests before.

westonruter commented:
This PR now makes it so that the `/amp/` endpoint is only presented as an available option if there is no existing post, term, etc that has already claimed the `amp` slug:

![image](https://user-images.githubusercontent.com/134745/99764642-5bf10700-2ab2-11eb-8cd0-37ad6a332058.png)

westonruter committed:
Add filtering for wp_unique_post_slug and wp_unique_term_slug

westonruter committed:
Remove filter_unique_term_slug since only applies when duplicate

westonruter committed:
Flush rewrite rules when swiching template mode

westonruter committed:
Restore environment variables after modifying for parse_request

westonruter committed:
Extract has/remove logic into methods

westonruter committed:
Move redirection logic to PairedAmpRouting service

Ensure rewrite rules are always updated.

westonruter committed:
Remove redundant is_validate_request check in amp_is_request()

westonruter committed:
Factor out duplicate redirect logic

westonruter committed:
Improve identification of comment form in Comment sanitizer

westonruter committed:
Remove redundant paired endpoint being added to _wp_amp_action_xhr_converted POST forms

westonruter committed:
Fix AMP-to-AMP linking for forms with implicit GET method

westonruter committed:
Move paired comment form handling from AMP_Theme_Support to MobileRedirection service

* Replace AMP_Theme_Support::amend_comment_form() with filtering comment_post_redirect in MobileRedirection.
* Replace AMP_Theme_Support::amend_comments_link() with legacy-conditional add_noamp_mobile_query_var in MobileRedirection service.
* Prevent initializing MobileRedirection hooks when in Standard mode.
* Add int to phpdoc for $post arg passed to amp_is_post_supported().

westonruter committed:
Remove obsolete finish_init tests for logic moved to PairedAmpRouting

westonruter committed:
Deprecate amp_add_frontend_actions(), move logic to PairedAmpRouting

westonruter committed:
Improve AMP/non-AMP URL determination in add_admin_bar_menu_items

westonruter committed:
Improve conditional logic in amp_add_admin_bar_view_link()

westonruter committed:
Prevent legacy reader paired URL structure from being available when slug used; add notice

westonruter committed:
Only remove paired endpoint on non-canonical sites during normalization

westonruter committed:
Only consider endpoint suffix if paired URL structure uses it

westonruter committed:
Eliminate attempting to determine the semantic canonical URL

When needing to provide a link[rel=canonical], just provide the self URL on an AMP-first site. On a paired AMP site, simply provide the corresponding non-AMP URL for the current paired URL.

westonruter committed:
Rework redirect_extraneous_endpoint_suffix logic to account for any extraneous endpoint

* Redirect to remove query var on standard mode.
* Redirect to remove duplicate endpoints on paired AMP configurations (e.g. query var and path suffix).

westonruter committed:
Only add rewrite endpoint if not Standard mode

westonruter committed:
Ensure /amp/ endpoint suffix is present on $wp->request

westonruter committed:
Avoid removing rewrite endpoint when custom paired structure is in place

westonruter committed:
Disable use of suffix endpoints if not using permalinks

westonruter committed:
Tighten up method names

westonruter committed:
Move filter_query_vars to add_paired_hooks

westonruter committed:
Move Admin\PairedBrowsing to Paired\BrowsingApp

westonruter committed:
Revert "Move Admin\PairedBrowsing to Paired\BrowsingApp"

This reverts commit 71b1e3dd1aeb7134dd6122d62357b1a63766ae92.

westonruter committed:
Utilize PairedAmpRouting service in PairedBrowsing

westonruter committed:
Remove redundant paired from method names

westonruter committed:
Let endpoint suffix/query_var helper methods be static

westonruter committed:
Introduce classes for defining paired URL structures

westonruter committed:
Ignore phpstan error regarding instantiation of abstract class

westonruter committed:
Check paired URL structure if AMP query var is not present

westonruter committed:
Fix passing back custom structure in REST API

westonruter committed:
Omit value from query var in LegacyReaderUrlStructure

westonruter committed:
Identify paired request in environment at plugins_loaded

* Have all paired URL structures go through the same environment detection logic.
* Defer initializing PluginSuppression until plugins_loaded to allow custom paired URL structures to be used.
* Ensure that all paired URL structures work even when Reader theme is used.
* Eliminate use of rewrite endpoints altogether, even in legacy Reader mode. This removes the need for Activateable and Deactivateable logic.
* Use dependency injection of PairedRouting in PluginSuppression and ReaderThemeLoader services.

westonruter committed:
Add default implementation of has_endpoint

westonruter committed:
Restore allowing query var alone to enable AMP; fix tests

westonruter committed:
Fix root logic for restore_path_endpoint_in_environment

westonruter committed:
Move static methods into PairedUrls service

westonruter committed:
Make PairedUrls service singular

westonruter committed:
Instantiate PairedUrlStructure via injector

westonruter committed:
Rename 'suffix endpoint' to 'path suffix' label

westonruter committed:
Rename 'endpoint suffix' with 'path suffix'

westonruter committed:
Check for not empty instead of isset to fix compat with PHP8

westonruter committed:
Guard against URL parse failure in add_path_suffix

westonruter committed:
Prevent registering AMP preview button plugin on standard mode

westonruter committed:
Fix phpunit tests after f8f30224

schlessera commented on assets/src/settings-page/paired-url-structure.js:
```diff
@@ -0,0 +1,258 @@
+ * @return {string} Sources string.
+ */
+function getCustomPairedStructureSources( sources ) {
+	let message = __( 'The custom structure is being introduced by:', 'amp' ) + ' ';
```
think the localized strings here will break on LTR because of the concatenation. A single translatable string with multiple placeholders should be used.

westonruter replied:
That's not the case, actually. The [Unicode Bidirectional Algorithm](https://www.w3.org/International/articles/inline-bidi-markup/uba-basics.en) differentiates between the [visual vs. logical ordering of text](https://www.w3.org/International/questions/qa-visual-vs-logical). So the whitespace here comes logically “after” the translated text. When the RTL direction is active, then “after” will be visually displayed on the left side of the text, whereas when the LTR direction is active, the “after” is on the right side of the text (matching the visual ordering of this LTR source code). Here's an example: https://codepen.io/westonruter/pen/rNMegvZ

schlessera commented on assets/src/settings-page/paired-url-structure.js:
```diff
@@ -0,0 +1,258 @@
+						disabled={ isCustom }
+					/>
+					<label htmlFor="paired_url_structure_query_var">
+						{ __( 'Query parameter', 'amp' ) + ': ' }
```
The punctuation should probably be part of the translation here, to avoid issues with LTR.

schlessera replied:
There are more instances below...

westonruter replied:
See https://github.com/ampproject/amp-wp/pull/5558#discussion_r537237382

schlessera commented on includes/validation/class-amp-validated-url-post-type.php:
```diff
@@ -2199,41 +2201,42 @@ public static function print_status_meta_box( $post ) {
+						 */
+						$actions = apply_filters( 'amp_validated_url_status_actions', $actions, $post );
+
+						echo implode( ' | ', array_filter( $actions ) ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Shouldn't `$actions` be run through `wp_kses` here?

westonruter replied:
Yeah, why not. Done in 1835abc2f.

schlessera commented on src/Admin/PairedBrowsing.php:
```diff
@@ -0,0 +1,313 @@
+		);
+		add_filter(
+			'script_loader_tag',
+			static function( $tag, $script_handle ) use ( $dev_mode_handles ) {
```
```suggestion
			static function ( $tag, $script_handle ) use ( $dev_mode_handles ) {
```

PHPCS should actually catch that one ...

schlessera commented on src/Admin/PairedBrowsing.php:
```diff
@@ -0,0 +1,313 @@
+	/**
+	 * Remove any unnecessary query vars that could hamper the paired browsing experience.
+	 *
+	 * @return bool Whether redirection was done.
```
This seems to be `true` when redirection was needed but actually could not be done (probably because of already sent headers...).

westonruter replied:
Clarified in ac24eca22.

schlessera commented on src/OptionsRESTController.php:
```diff
@@ -194,6 +194,20 @@ static function( $slug ) {
+		 *
+		 * @param array $service_options REST Options for Services.
+		 */
+		$service_options = apply_filters( 'amp_rest_options', [] );
```
Filters should always use extra hardening:
```suggestion
		$service_options = array_filter( (array) apply_filters( 'amp_rest_options', [] ) );
```

schlessera commented on src/OptionsRESTController.php:
```diff
@@ -338,6 +352,20 @@ public function get_item_schema() {
+			 *
+			 * @param array $schema Schema.
+			 */
+			$services_schema = apply_filters( 'amp_rest_options_schema', [] );
```
Filters should always use extra hardening:
```suggestion
			$services_schema = array_filter( (array) apply_filters( 'amp_rest_options_schema', [] ) );
```

schlessera commented on src/PairedRouting.php:
```diff
@@ -0,0 +1,937 @@
+	 * @see WP_Rewrite::using_permalinks()
+	 * @var string
+	 */
+	const ENDPOINT_USING_PERMALINKS = 'using_permalinks';
```
These should probably both either use or not use the `endpoint_` prefix.

westonruter replied:
Addressed by changing this instance to use `REWRITE_` as a prefix which is more appropriate. See a9b68cd857971f671555aa0db4ec4a8b91fb34f2 given that it reflects a property in `WP_Rewrite`.

schlessera commented on src/PairedRouting.php:
```diff
@@ -0,0 +1,937 @@
+	/**
+	 * Determine a given URL is for a paired AMP request.
+	 *
+	 * If no URL was is provided, then it will check whether WordPress has already parsed the AMP
```
```suggestion
	 * If no URL is provided, then it will check whether WordPress has already parsed the AMP
```

schlessera commented on src/PairedRouting.php:
```diff
@@ -0,0 +1,937 @@
+	 * Determine a given URL is for a paired AMP request.
+	 *
+	 * If no URL was is provided, then it will check whether WordPress has already parsed the AMP
+	 * query var as part of the request. If still not present, then it will use get the current URL
```
```suggestion
	 * query var as part of the request. If still not present, then it will get the current URL
```

schlessera commented on src/PairedUrl.php:
```diff
@@ -0,0 +1,146 @@
+<?php
+/**
+ * Class PairedUrls.
```
```suggestion
 * Class PairedUrl.
```

schlessera commented on src/PairedUrlStructure.php:
```diff
@@ -0,0 +1,59 @@
+	 * @param string $url URL (or REQUEST_URI).
+	 * @return bool True if the URL has the paired endpoint.
+	 */
+	public function has_endpoint( $url ) {
+		return $url !== $this->remove_endpoint( $url );
+	}
```
This logic is never used as it is overridden in all 4 implementations. We should make this abstract as well.
```suggestion
	abstract public function has_endpoint( $url );
```

westonruter replied:
See examples in custom paired URL structures:

* https://gist.github.com/westonruter/67487d199f4cbc1968a30c60de121893#file-subdomainurlstructure-php
* https://gist.github.com/westonruter/f881658cb1bc5a03cc432c8e2b89102a#file-prefixurlstructure-php

I haven't removed the `has_endoint` method from the bundled paired URL structures yet since I wasn't sure yet if there's some value in having tailored `has_endpoint` implementations.

schlessera commented on src/PairedUrlStructure/LegacyTransitionalUrlStructure.php:
```diff
@@ -0,0 +1,50 @@
+use AmpProject\AmpWP\PairedUrlStructure;
+
+/**
+ * Descriptor for paired URL structures that end in /amp/ path suffix for non-hierarchical posts and ?amp for others.
```
This doesn't look right base don the code.

westonruter replied:
Correct. I copied it from `LegacyReaderUrlStructure`. Copy pasta fixed in e32401749.

schlessera commented on src/PairedUrl.php:
```diff
@@ -0,0 +1,146 @@
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.1
+ * @internal
```
This class is needed as a constructor argument when creating a custom `PairedUrlStructure` implementation, so it shouldn't be marked as `@internal`.
```suggestion
```

schlessera commented:
For me the CSS is not being built correctly:
<img width="869" alt="AMP Paired Browsing_ WordPress Develop – Just another WordPress site - Goog  2020-12-05 at 6 20 45 PM    WhatsApp - Google Chrome" src="https://user-images.githubusercontent.com/83631/101260285-20e80800-372f-11eb-9777-865f5bb7adf9.png">

westonruter committed:
Remove erroneous internal tag from PairedUrl service

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Fix class name in phpdoc for PairedUrl service

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Fix grammar typo

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Harden with array casting after applying filters

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Fix grammar typo

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Improve whitespace for closure

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Rename using_permalinks to rewrite_using_permalinks

westonruter committed:
Use alternative array check to prevent filtering out empty values

westonruter committed:
Add REWRITE_USING_PERMALINKS to REST API schema

westonruter committed:
Utilize wp_kses for hardening amp_validated_url_status_actions

westonruter committed:
Clarify behavior of ensure_app_location

westonruter committed:
Fix description of LegacyTransitionalUrlStructure

westonruter committed:
Restore paired-browsing app.css internal dependency

westonruter commented:
> For me the CSS is not being built correctly

Oops. Fixed in 1d76c79ac.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/amp-url-customization

* 'develop' of github.com:ampproject/amp-wp: (117 commits)
  Use codeCoverageIgnore
  Make sure menu button container remains visible
  Persist focus on mobile menu button when toggling
  Unset wp_scripts and wp_styles after each test
  Add test for amend_twentytwentyone_styles
  Add test for amend_twentytwentyone_dark_mode_styles
  Remove !important rule
  Trap keyboard navigation within menu when it's open
  Replace amp-lightbox with an amp-bind solution
  Bump sirbrillig/phpcs-variable-analysis from 2.10.0 to 2.10.1
  Improve diff output for PHP files
  Try to maintain focus on button when opening and closing lightbox
  Use DOMElement instanceof check for static analysis
  Apply padding to spans within menu button toggle
  Use less specific XPath query to get menu toggles
  Add a warning notice to setting
  Make phpstan happy
  Use two discrete toggles for opening and closing mobile menu in Twenty Twenty-One
  Add a warning notice to setting
  Add force flag to rm on the productionVendorExcludedFilePatterns
  ...

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/amp-url-customization

* 'develop' of github.com:ampproject/amp-wp: (298 commits)
  Further refine messaging related to validation issues
  Prevent invalid-block-outline from causing block to be unclickable
  Expand all validation errors associatd with a block and ensure in viewport
  Add missing class to amp-toolbar-icon
  Simplify logic to ensure -1 <= limit_per_type <= PHP_INT_MAX
  Ensure amp_url_validation_limit_per_type filter returns int
  Mark URLValidationProvider and ScannableURLProvider as internal
  Guard against empty post ID causing global post to be returned by get_post()
  Make RecurringBackgroundTask::schedule_event() final
  Fix ability for BackgroundTaskDeactivator to unschedule hooks
  Reduce URLValidationCron frequency to daily
  Remove obsolete test assertion
  AMP block toolbar button: prevent closing error if already opened
  Make CronBasedBackgroundTask non-Conditional
  Update service name for SavePostValidationEvent and add to PHPSTORM_META
  Add missing $args to wp_next_scheduled() call after 1419ef4
  Switch to google-github-actions/setup-gcloud action
  Only run PHPStan if there is a modified PHP file
  Run composer update
  Ensure alphabetical order for patches in composer.json
  ...

westonruter committed:
Add caching for url_to_postid() and address phpcs complains from WordPress-VIP

westonruter committed:
Fix eslint issues identified by #5732

westonruter committed:
Ensure wp_old_slug_redirect() is called just-in-time before redirect_paired_amp_unavailable

Fixes #3357

westonruter committed:
Remove assertions due to amp_add_frontend_actions() no longer being called in AMP_Theme_Support

westonruter commented on src/MobileRedirection.php:
```diff
@@ -406,6 +431,33 @@ public function add_mobile_alternative_link() {
 		);
 	}
 
+	/**
+	 * Redirect to AMP page after submitting comment if the URL is on this site.
+	 *
+	 * @param string $url URL.
+	 * @return string Amended URL.
+	 */
+	public function filter_comment_post_redirect( $url ) {
+		if (
+			isset( AMP_HTTP::$purged_amp_query_vars[ AMP_HTTP::ACTION_XHR_CONVERTED_QUERY_VAR ] )
+			&&
+			wp_parse_url( home_url(), PHP_URL_HOST ) === wp_parse_url( $url, PHP_URL_HOST )
+		) {
+			$url = $this->paired_routing->add_endpoint( $url );
+		}
+		return $url;
+	}
```
I need to double-check why this was added as I don't see it being necessary.

westonruter replied:
Yes, it is needed. a31e34887

westonruter committed:
Add missing type definitions to JS

westonruter committed:
Add tooltip to top-level AMP admin bar menu item

westonruter committed:
Revert changes to is_paired_available but leave soft-deprecation

westonruter committed:
Add comment to explain need for filter_comment_post_redirect

westonruter commented:
@pierlon This is ready for review. Sorry it's so massive. I tried to detail all of the changes in the description.

While you review I'll be working on adding tests for the new services.

westonruter committed:
Add PHPUnit tests for PairedBrowsing

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/amp-url-customization

* 'develop' of github.com:ampproject/amp-wp:
  Make the E2E job required
  Set default timeout to 10s
  Use new arg format for `deleteTheme()`
  Remove E2E test utils that are now available upstream

westonruter committed:
Set Transitional mode in test_add_admin_bar_menu_item

westonruter committed:
Move PairedBrowsingTest to Admin directory

westonruter committed:
Add incomplete tests for PairedRouting

westonruter committed:
Revert "Add incomplete tests for PairedRouting"

This reverts commit d122081315a9ac1406f328c016df2b0358e20ef5.

westonruter committed:
Revert "Move PairedBrowsingTest to Admin directory"

This reverts commit c26bad48719598aea3f8574b9bc77f495572e154.

westonruter committed:
Revert "Revert "Move PairedBrowsingTest to Admin directory""

This reverts commit bc91570a295dbbe8c48683e354e913cfd40cfe67.

westonruter committed:
Update namespace for PairedBrowsingTest to hopefully fix strange test failures in WP≤5.1

westonruter commented:
So confused. In bc91570 there are no PHPUnit test failures:

![image](https://user-images.githubusercontent.com/134745/103401758-1b4bcb00-4aff-11eb-95d7-c6df17d3efa3.png)

 If I then move `PairedBrowsingTest` into the `Admin` directory (d046b27 and 2e62da4), then there are test failures exclusively in WP  4.9, 5.0, and 5.1:

![image](https://user-images.githubusercontent.com/134745/103401777-2c94d780-4aff-11eb-9058-1510dd4958c4.png)

![image](https://user-images.githubusercontent.com/134745/103401806-43d3c500-4aff-11eb-8366-d85b880ce363.png)

I'm baffled because these are the only changes: https://github.com/ampproject/amp-wp/compare/bc91570...2e62da4

westonruter commented:
And why would the tests pass in other PHP/WP versions?

westonruter commented:
- [x] Todo: Re-revert 577454b.

pierlon commented:
> And why would the tests pass in other PHP/WP versions?

Hmm, taking a look.

westonruter committed:
Revert "Revert "Add incomplete tests for PairedRouting""

This reverts commit 577454b7b6fab6b79cb032d8954e041cc04eece8.

westonruter committed:
Check query_var for post types and taxonomies in addition to rewrite slug

westonruter committed:
Update cached url_to_postid() variant to not short-circuit if host not supplied

westonruter committed:
Add integrated test for initialize_paired_request

westonruter committed:
Improve method ordering

westonruter committed:
Test REQUEST_URI suspension duriong parse request

westonruter committed:
Add tests for filter_unique_post_slug, add_paired_request_hooks, get_wp_rewrite

westonruter committed:
Add tests for filter_default_options and sanitize_options

westonruter committed:
Add tests for has/add/remove_endpoint, has_custom_paired_url_structure, get_all_structure_paired_urls, get_paired_url_examples, get_custom_paired_structure_sources, correct_query_when_is_front_page, maybe_add_paired_endpoint

westonruter committed:
Remove obsolete rewrite endpoint condition logic from redirect_extraneous_paired_endpoint

Also eliminate unnecessary (and perhaps over-eager) removal of AMP query var during a 404 when a suffix is present.

Add tests for redirect_extraneous_paired_endpoint.

westonruter committed:
Refactor redirect_paired_amp_unavailable to dedup with redirect_paired_amp_unavailable

Remove obsolete rewrite endpoint condition logic from redirect_extraneous_paired_endpoint

Finish test coverage of PairedRouting

westonruter committed:
Add tests for PairedUrl

westonruter committed:
Add tests for PairedUrlStructures

westonruter commented:
PHPUnit test coverage should now be complete.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/amp-url-customization

* 'develop' of github.com:ampproject/amp-wp: (33 commits)
  Update docs
  Bump stable tag to 2.0.10
  Add tests
  Restore intrinsic layout to amp-video in amp-story-page-attachment
  Use nonce scheme instead of auth scheme
  Utilize hash_equals() in unserialize_validation_error_messages()
  Fix jsdoc type as ReactElement
  Persist URL fragment through client-side mobile redirection
  Bump postcss from 8.2.2 to 8.2.4
  Bump eslint-plugin-jsdoc from 30.7.13 to 31.0.3
  Prevent infinite mobile redirect when non-AMP page served as AMP
  Remove the `class` attribute from `amp-fit-text` blocks to unblock block deprecation
  Update AMP validator spec to 8e02747, restoring intrinsic layout to amp-video
  Update validator spec to 1130
  Update amphtml spec to 2011252111002
  Bump @wordpress/compose from 3.23.0 to 3.23.1
  Bump @wordpress/e2e-test-utils from 4.16.0 to 4.16.1
  Bump @wordpress/edit-post from 3.26.0 to 3.26.1
  Bump @wordpress/url from 2.20.0 to 2.21.0
  Bump @wordpress/plugins from 2.24.0 to 2.24.1
  ...

pierlon commented on assets/src/admin/paired-browsing/app.js:
```diff
@@ -46,47 +94,127 @@ class PairedBrowsingApp {
+			return;
+		}
+
+		if ( ! this.isAmpWindow( event.source ) && ! this.isNonAmpWindow( event.source ) ) {
```
As an added precaution, I'd recommend verifying that the `event.origin` matches the URL origin of the non-AMP or AMP iframe.

Something like:

```js
if ( ! [ currentAmpOrigin, currentNonAmpOrigin ].includes( event.origin ) ) {
	return;
}
```

(where `currentAmpOrigin` and `currentNonAmpOrigin` are to be determined of course.)

westonruter replied:
Good. This closes the loop with the `origin` restriction coming from the client:

https://github.com/ampproject/amp-wp/blob/891847a4f4d2ab006e986f9361ec3d751f2dfb0e/assets/src/admin/paired-browsing/client.js#L44-L53

westonruter replied:
Added in a59a21e18762376ce6a401c6b0c7a82a7dce277d

pierlon commented on assets/src/admin/paired-browsing/app.js:
```diff
@@ -227,108 +366,112 @@ class PairedBrowsingApp {
-	 * @param {Window} win Document window.
+	 * @param {Object}  data                   Data.
+	 * @param {boolean} data.isAmpDocument     Whether the document is actually an AMP page.
+	 * @param {?string} data.ampUrl            The AMP URL.
+	 * @param {?string} data.nonAmpUrl         The non-AMP URL.
```
Is it the case that `data.ampUrl` or `data.nonAmpUrl` are not required?

westonruter replied:
You're right, they're always provided:

https://github.com/ampproject/amp-wp/blob/891847a4f4d2ab006e986f9361ec3d751f2dfb0e/assets/src/admin/paired-browsing/client.js#L137-L148

This may have come from before, as I recall I was passing one or the other.

westonruter replied:
Fixed by you in 2be5e09f404838ad31d2210d2e119eea344c7882

pierlon commented on assets/src/admin/paired-browsing/app.js:
```diff
@@ -227,108 +366,112 @@ class PairedBrowsingApp {
+	 * @param {?string} data.ampUrl            The AMP URL.
+	 * @param {?string} data.nonAmpUrl         The non-AMP URL.
+	 * @param {string}  data.documentTitle The canonical link URL if present.
+	 * @param {Window}  sourceWindow The source window.
```
It seems `data.ampUrl` and `data.nonAmpUrl` are required, so I'd suggest removing the `?` in the param type. Also, it seems the description for `documentTitle` is incorrect, so I'm suggesting a correct one.

```suggestion
	 * @param {Object}  data                Data.
	 * @param {boolean} data.isAmpDocument  Whether the document is actually an AMP page.
	 * @param {string} data.ampUrl         The AMP URL.
	 * @param {string} data.nonAmpUrl      The non-AMP URL.
	 * @param {string}  data.documentTitle  The title of the document.
	 * @param {Window}  sourceWindow        The source window.
```

hasanbahtiar commented:
Hi AMP Plugin Team, I use category as subdomain plugins, this plugins work on AMP Transitional with endpoint ?amp, but i want change amp url to path prefix.

I download and install plugin in github only work on main domain, not working on subdomain category, but canonical true i view on view-source.

Please help Mister.

westonruter commented:
@hasanbahtiar I'm not entirely sure I understand what you're use case is, but if you're wanting AMP URLs to look like `amp.example.com` instead of `example.com/?amp` then you may consider an extension plugin which implements a custom paired URL structure for subdomains: https://gist.github.com/westonruter/67487d199f4cbc1968a30c60de121893

Nevertheless, it is possible that this will need to be modified to fit your site's setup.

We won't be officially supporting the use of subdomains for paired URLs, but we will provide the extensibility so you can implement your own scheme. The above subdomain paired URL structure plugin is intended to give you a starting point to construct paired AMP URLs as you need.

pierlon commented on assets/src/admin/paired-browsing/client.js:
```diff
@@ -3,33 +3,169 @@
+ * @param {MessageEvent} event
+ */
+function receiveMessage( event ) {
+	if ( ! event.data || ! event.data.ampPairedBrowsing || ! event.data.type || ! event.source ) {
```
Similar to above, verifying that `event.origin` is equal to `nonAmpUrlObject.href` (or `parent.location.origin`) would be a great check to add.

westonruter replied:
Done in a59a21e18762376ce6a401c6b0c7a82a7dce277d

westonruter replied:
(We wouldn't be able to use `parent.location.origin` in a cross-domain context.)

pierlon commented on assets/src/admin/paired-browsing/app.css:
```diff
@@ -189,7 +192,3 @@ iframe {
 	width: auto;
 }
-
-.hidden {
-	display: none;
-}
```
Was this intended to be removed? I see that the class is still being referred to at:

https://github.com/ampproject/amp-wp/blob/1076da51e250eec604db07670822e2fc69ab0d14/assets/src/admin/paired-browsing/app.js#L306

westonruter replied:
Oh yeah, I think I meant to switch this to use the `hidden` property instead of the `hidden` class. Updated in 73b8600bf.

hasanbahtiar commented:
I like this, path prefix (e.g. https://example.com/amp/foo/) not subdomain (e.g. https://amp.example.com/foo/) mister.

I install plugins, work only main domain www.example.com/amp/post but not working on subcat.example.com/amp/post.

Of Transitional normaly, www.example.com/post/?amp and in subdoman.example.com/post/?amp work. But i want change to path prefix (e.g. https://example.com/amp/foo/).

I use plugins https://github.com/wp-plugins/main-category-as-subdomain

Help me mister!

westonruter commented:
A plugin which implements AMP URLs via a path prefix is: https://gist.github.com/westonruter/f881658cb1bc5a03cc432c8e2b89102a

You should be able to adapt it for your needs. 

All such plugins are prototypes and not officially supported. The `PairedUrlStructure` abstract class is provided for you to implement your own schemes for paired URL structures.

westonruter committed:
Eliminate reliance on WP_Rewrite in favor of looking at permalink_structure option

pierlon commented on assets/src/settings-page/paired-url-structure.js:
```diff
@@ -0,0 +1,264 @@
+							)
+						}
+					</p>
+					<PairedUrlExamples pairedUrls={ editedOptions.paired_url_examples.custom } />
```
I suppose this was copypasta from above.

```suggestion
```

westonruter replied:
Good catch. Indeed.

pierlon commented on includes/class-amp-theme-support.php:
```diff
@@ -380,87 +353,17 @@ static function() {
 	 *
 	 * @since 0.7
+	 * @deprecated No longer used. Consider instead `! amp_is_canonical() && amp_is_available()`.
+	 * @todo There are ecosystem plugins which are still using this method. See <https://wpdirectory.net/search/01EPD9M5CKWHJ7NMQ1WE0YGPJ5>.
```
Is `_deprecated_function()` to be used here?

westonruter replied:
Not yet. I'd want to reach out to the plugin authors in the directory first to eliminate it from being used: https://wpdirectory.net/search/01EPD9M5CKWHJ7NMQ1WE0YGPJ5

pierlon commented on src/Admin/PairedBrowsing.php:
```diff
@@ -0,0 +1,322 @@
+/**
+ * Managing the paired browsing app.
+ *
+ * @since 1.5.0
```
```suggestion
 * @since 2.1
```

pierlon commented on src/Admin/PairedBrowsing.php:
```diff
@@ -0,0 +1,322 @@
+	 *
+	 * @var UserAccess
+	 */
+	public $dev_tools_user_access;
+
+	/**
+	 * Paired Routing.
+	 *
+	 * @var PairedRouting
+	 */
+	public $paired_routing;
```
Do these need to be public?

westonruter replied:
No, but the whole class is `@internal` so there doesn't seem to be a need to make them `private`. It makes them easier to access in unit testing.

pierlon commented on src/Admin/PairedBrowsing.php:
```diff
@@ -0,0 +1,322 @@
+		if ( wp_safe_redirect( $updated_url ) ) {
+			// @codeCoverageIgnoreStart
+			exit;
+			// @codeCoverageIgnoreEnd
```
```suggestion
			exit; // @codeCoverageIgnore
```

pierlon commented on src/MobileRedirection.php:
```diff
@@ -406,6 +431,35 @@ public function add_mobile_alternative_link() {
+		if (
+			isset( AMP_HTTP::$purged_amp_query_vars[ AMP_HTTP::ACTION_XHR_CONVERTED_QUERY_VAR ] )
+			&&
+			wp_parse_url( home_url(), PHP_URL_HOST ) === wp_parse_url( $url, PHP_URL_HOST )
```
Just a note: Users who opt for the paired subdomain URL solution would need to ensure that `home_url()` returns the correct value here to ensure the condition holds true.

pierlon commented on src/Option.php:
```diff
@@ -43,6 +43,53 @@ interface Option {
+	 *
+	 * @var string
+	 */
+	const PAIRED_URL_STRUCTURE_QUERY_VAR = 'query_var';
+
+	/**
+	 * Path suffix paired URL structure.
+	 *
+	 * This adds `/amp/` to all URLs, even pages and archives. This is a popular option for those who feel query params
+	 * are bad for SEO.
+	 *
+	 * @var string
+	 */
+	const PAIRED_URL_STRUCTURE_PATH_SUFFIX = 'path_suffix';
+
+	/**
+	 * Legacy transitional paired URL structure.
+	 *
+	 * This involves using `?amp` for all paired AMP URLs.
+	 *
+	 * @var string
+	 */
+	const PAIRED_URL_STRUCTURE_LEGACY_TRANSITIONAL = 'legacy_transitional';
+
+	/**
+	 * Legacy transitional paired URL structure.
+	 *
+	 * This involves using `/amp/` for all non-hierarchical post URLs which lack endpoints or query vars, or else using
+	 * the same `?amp` as used by legacy transitional.
+	 *
+	 * @var string
+	 */
+	const PAIRED_URL_STRUCTURE_LEGACY_READER = 'legacy_reader';
```
Wouldn't these be better suited for `\AmpProject\AmpWP\PairedRouting::class`?

pierlon replied:
Speaking of, the `PAIRED_URL_STRUCTURE_CUSTOM` constant is in that class instead of this one.

westonruter replied:
I was somewhat torn there. The reason why I put `PAIRED_URL_STRUCTURE_CUSTOM` into `PairedRouting` is that the value is never used in a saved option. So it is used exclusively for a value in the REST API as opposed a value in an option. This is in contrast with `legacy_reader` which is persisted to the DB.

pierlon replied:
I see, that makes sense :+1:.

pierlon commented on src/PairedRouting.php:
```diff
@@ -0,0 +1,901 @@
+	/**
+	 * Sanitize options.
+	 *
+	 * @todo This is redundant with the enum defined in the schema.
```
How so? Wouldn't this ensure that the option is sanitized correctly when the `amp_options` option is being updated in a non-REST manner?

westonruter replied:
Correct, but as I recall we exclusively use the REST API to update settings. Nevertheless, I'll change it as follows:

```suggestion
	 * Note that in a REST API context this is redundant with the enum defined in the schema.
```

pierlon commented on src/PairedRouting.php:
```diff
@@ -0,0 +1,901 @@
+			if ( ! $structure_class || ! is_subclass_of( $structure_class, PairedUrlStructure::class ) ) {
+				$structure_slug = AMP_Options_Manager::get_option( Option::PAIRED_URL_STRUCTURE );
+				if ( array_key_exists( $structure_slug, static::PAIRED_URL_STRUCTURES ) ) {
+					$structure_class = static::PAIRED_URL_STRUCTURES[ $structure_slug ];
```
```suggestion
				if ( array_key_exists( $structure_slug, self::PAIRED_URL_STRUCTURES ) ) {
					$structure_class = self::PAIRED_URL_STRUCTURES[ $structure_slug ];
```

pierlon commented on src/PairedRouting.php:
```diff
@@ -0,0 +1,901 @@
+				]
+			);
+			foreach ( $posts as $post ) {
+				if ( count( AMP_Post_Type_Support::get_support_errors( $post ) ) !== 0 ) {
```
Just a note: No examples would be shown if there are no published posts for the queried post types, or if all posts for a certain post type did not support AMP (both scenarios are unlikely, but possible).

westonruter replied:
True. Worst case is just no examples would be shown:

![image](https://user-images.githubusercontent.com/134745/105307407-52972000-5b71-11eb-8830-816d5c028026.png)

But the examples are already somewhat redundant since the forms are indicated in the options already.

pierlon commented on src/PairedUrl.php:
```diff
@@ -0,0 +1,146 @@
+<?php
```
This more looks like a helper trait than a service, so I'd recommend making it one.

westonruter replied:
Or else it the methods in that service could be just made methods on the abstract class directly. 

The reason a service was chosen was based on the advice of @schlessera. 

Also, the service is being used not only in the `PairedUrlStructure` subclasses but also in the `PairedRouting` service.

pierlon commented on src/PairedUrlStructure.php:
```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * Abstract class PairedUrl.
```
```suggestion
 * Abstract class PairedUrlStructure.
```

pierlon commented on src/PairedUrlStructure/QueryVarUrlStructure.php:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Descriptor for paired URL structures that include the `?amp=1` query parameter.
+ *
+ * @todo Rename to QueryArg for better parity with add_query_arg(). Or QueryParam parameter for better spec alignment?
```
I'd go with `QueryParam`.

westonruter replied:
Humm. I just went through renaming it in d29f652de8cc6752d64fbda865ff5d9c13326ce2, but then I'm seeing references to `query_var` all over the place for this idea, including the `QueryVar` interface. Would not `query_var` be more consistent in the codebase instead of `query_param`?

pierlon replied:
Oh I didn't know you meant for it to be such as extensive modification; I thought it was limited to the name of the class (and probably what's within it, and its references).

I think using `query_var` would work just fine. As long as a comment or description is added in places where uncertainty is bound I don't see any issues with it.

westonruter replied:
Reverted in 651b41155a0078fb4afcc1c7ac44b582dd9c430b

pierlon commented on src/PluginSuppression.php:
```diff
@@ -87,13 +96,22 @@ function ( $props, $block_name ) {
 		// could be done early for Transitional mode as well since a query parameter is also used for frontend requests
 		// but there is no similar need to suppress the registration of Customizer controls in Transitional mode since
 		// there is no separate Customizer for AMP in Transitional mode (or legacy Reader mode).
+		// @todo This could rather check if ! amp_is_canonical().
```
You mean to replace the entire condition with `! amp_is_canonical()`?

westonruter replied:
Yes. Or rather, it could be:

```suggestion
		// @todo This check could be replaced with ( ! amp_is_canonical() && $this->paired_routing->has_endpoint() ).
```

pierlon replied:
Yea that could work.

pierlon reviewed (changes requested):
Sorry for the delay in reviewing this. I spent a good amount of time trying to break this but there was always some sort of sanity check somewhere that would catch the would-be bug :smile:. Nevertheless, there are some suggestions I'd like to recommend.

westonruter committed:
Add origin checks for message events

westonruter committed:
Remove nullable from jsdoc args and fix param description

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Make use of codeCoverageIgnore comment

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Fix class reference

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Fix since phpdoc param

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Remove PairedUrlExamples copy pasta

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Use self instead of static

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

pierlon commented on tests/php/src/Admin/PairedBrowsingTest.php:
```diff
@@ -0,0 +1,244 @@
+		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::STANDARD_MODE_SLUG );
+		$this->assertFalse( PairedBrowsing::is_needed() );
+
+		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::TRANSITIONAL_MODE_SLUG );
```
Maybe add an assertion for `AMP_Theme_Support::READER_MODE_SLUG` just for completeness: 

```suggestion
		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::READER_MODE_SLUG );
		$this->assertFalse( PairedBrowsing::is_needed() );

				AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::TRANSITIONAL_MODE_SLUG );
```

pierlon commented on tests/php/src/PairedUrlStructure/LegacyTransitionalUrlStructureTest.php:
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace AmpProject\AmpWP\Tests;
```
```suggestion
namespace AmpProject\AmpWP\Tests\PairedUrlStructure;
```

pierlon commented on tests/php/src/PairedUrlStructure/LegacyReaderUrlStructureTest.php:
```diff
@@ -0,0 +1,188 @@
+<?php
+
+namespace AmpProject\AmpWP\Tests;
```
```suggestion
namespace AmpProject\AmpWP\Tests\PairedUrlStructure;
```

pierlon commented on tests/php/src/PairedUrlStructure/PathSuffixUrlStructureTest.php:
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace AmpProject\AmpWP\Tests;
```
```suggestion
namespace AmpProject\AmpWP\Tests\PairedUrlStructure;
```

pierlon commented on tests/php/src/PairedUrlStructure/QueryVarUrlStructureTest.php:
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace AmpProject\AmpWP\Tests;
```
```suggestion
namespace AmpProject\AmpWP\Tests\PairedUrlStructure;
```

westonruter committed:
Change a todo into a note

westonruter committed:
Use hidden prop instead of hidden class

pierlon commented on tests/php/test-class-amp-options-manager.php:
```diff
@@ -9,13 +9,14 @@
  * @covers AMP_Options_Manager
  */
-class Test_AMP_Options_Manager extends WP_UnitTestCase {
+class Test_AMP_Options_Manager extends DependencyInjectedTestCase {
```
`DependencyInjectedTestCase` doesn't seem to be required.

```suggestion
class Test_AMP_Options_Manager extends WP_UnitTestCase {
```

pierlon commented on tests/php/test-class-amp-options-manager.php:
```diff
@@ -9,13 +9,14 @@
 use AmpProject\AmpWP\Tests\Helpers\AssertContainsCompatibility;
 use AmpProject\AmpWP\Tests\Helpers\LoadsCoreThemes;
 use AmpProject\AmpWP\Admin\ReaderThemes;
+use AmpProject\AmpWP\Tests\DependencyInjectedTestCase;
```
Suggestion to go with above.

```suggestion
```

westonruter committed:
Rename query_var to query_param as a paired URL structure

westonruter committed:
Add assertion for completeness

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Fix test class namespaces

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Add newly-needed use statements for DependencyInjectedTestCase

westonruter committed:
Use WP_UnitTestCase for Test_AMP_Options_Manager

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Update todo for how a condition can be updated

westonruter committed:
Go back to using query_var instead of query_param

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/amp-url-customization

* 'develop' of github.com:ampproject/amp-wp:
  Use wp_robots() instead of noindex() in WP 5.7 (#5793)
  Bump eslint-plugin-jsdoc from 31.0.3 to 31.0.4
  Bump google/cloud-storage from 1.23.0 to 1.23.1
  Bump @wordpress/block-editor from 5.2.0 to 5.2.1

pierlon committed:
Ensure `$current_screen` global variable does not interfere with tests

pierlon committed:
Ensure "add analytics" button is in view

pierlon committed:
Wait for succss notice to appear

pierlon reviewed (approved):
A thing of beauty, well done! :shipit:

westonruter commented:
@pierlon Thank you for the thorough review and for getting the tests across the finish line. 🤝

westonruter commented:
Follow up PR to fix an issue where a URL with an empty query param (e.g. `/?foo=`) would get misidentified as an AMP URL: https://github.com/ampproject/amp-wp/pull/5804
